### DISCUSSION TITLE
Debugging a blank screen

### DISCUSSION THREAD

**Discussion by Luke-Shepp** [rsta2/circle#492] [Q&A]
I had a propect with circle, circle-stdlib and lvgl up and running on a pi zero 2w some time ago now, but abandoned it because the graphics were extremely laggy. I've recently got a Pi5 and have started porting it over. After some initial problems building for aarch64 and pi5, I can successfully build kernel images now.

The issue I'm facing is that the hdmi output is blank. The backlight is there, the screen just has nothing on it. No cursor, spinner etc.

I have tried building a few samples, and although sample 28 errors to say it can't get the touch buffer and the screen aspect ratio is squashed - it boots and shows log lines.

I've copied the kernel source over from that sample, to my project but still a blank screen. 

Something is clearly amiss with the rest of my code when running on a Pi5. Very odd that the same kernel.cpp/.h doesn't run in my project though.

Anyway, I can spend time investigating - but I'm not sure where to go from here. How would I debug a completely blank screen? Immediately after the logger is initialised I'm outputting a log message but it never shows up. 

For completeness, the kernel I'm using is below incase there's anything glaringly obvious, but I've stripped it well back to basically nothing in an attempt to get _something_ to show

```cpp
// kernel.cpp


#include "kernel.h"
#include <string>
#include <cstring>

static const char FromKernel[] = "kernel";

CKernel::CKernel (void)
:	m_Screen (1600, 600),
	m_Timer (&m_Interrupt),
	m_Logger (3, &m_Timer),
	m_USBHCI (&m_Interrupt, &m_Timer)
{
	//
}

CKernel::~CKernel (void)
{
}

boolean CKernel::Initialize (void)
{
	boolean bOK = TRUE;

	if (bOK)
	{
		bOK = m_Screen.Initialize ();
	}

	if (bOK)
	{
		bOK = m_Serial.Initialize (9600);
	}

	if (bOK)
	{
		CDevice *pTarget = m_DeviceNameService.GetDevice (m_Options.GetLogDevice (), FALSE);
		if (pTarget == 0)
		{
			pTarget = &m_Screen;
		}

		bOK = m_Logger.Initialize (pTarget);
	}

	m_Logger.Write(FromKernel, LogNotice, "Logger Init Done.");

	if (bOK)
	{
		bOK = m_Interrupt.Initialize ();
	}

	if (bOK)
	{
		bOK = m_Timer.Initialize ();
	}

	if (bOK)
	{
		bOK = m_USBHCI.Initialize ();
	}

	m_Logger.Write(FromKernel, LogNotice, "Kernel Init Done.");

	return bOK;
}

TShutdownMode CKernel::Run (void)
{
	m_Logger.Write (FromKernel, LogNotice, "Compile time: " __DATE__ " " __TIME__);

	for (unsigned nCount = 0; 1; nCount++)
	{
		m_Screen.Rotor(0, nCount);
		m_Timer.MsDelay(1000/60);
	}

	return ShutdownHalt;
}


```

```cpp
// kernel.h


#ifndef _kernel_h
#define _kernel_h

#include <circle/actled.h>
#include <circle/koptions.h>
#include <circle/devicenameservice.h>
#include <circle/screen.h>
#include <circle/serial.h>
#include <circle/exceptionhandler.h>
#include <circle/logger.h>
#include <circle/usb/usbhcidevice.h>
#include <circle/sched/scheduler.h>
#include <SDCard/emmc.h>
#include <fatfs/ff.h>
#include <circle/usertimer.h>

enum TShutdownMode
{
	ShutdownNone,
	ShutdownHalt,
	ShutdownReboot
};

class CKernel
{
public:
	CKernel (void);
	~CKernel (void);

	boolean Initialize (void);

	TShutdownMode Run (void);

private:
	// do not change this order
	CActLED				m_ActLED;
	CKernelOptions		m_Options;
	CDeviceNameService	m_DeviceNameService;
	CScreenDevice		m_Screen;
	CSerialDevice		m_Serial;
	CExceptionHandler	m_ExceptionHandler;
	CInterruptSystem	m_Interrupt;
	CTimer				m_Timer;
	CLogger				m_Logger;
	CScheduler			m_Scheduler;
	CUSBHCIDevice		m_USBHCI;
};

#endif

```

**Comment by rsta2** [comment]
The difference is the display support in the firmware of the RPi 5. It's not as comprehensive as the display support in the earlier models. I don't think a display resolution of 1600x600 will work with it. Have you tried `m_Screen(0, 0)`, when creating the screen driver object? This will automatically use the default resolution. I have tried your code with this, and it works with my Full HD display, while 1600x600 does not work.

For debugging without a working HDMI display, one can use an USB-to-serial adapter. BTW. The Official 7" Touchscreen can't be used on the RPi 5 with Circle. That's why the *sample/28-touchscreen* does not work.

**Reply by Luke-Shepp** [reply]
Unfortunately 0, 0 as the resolution has the same effect for me.

The screen is some Waveshare branded Aliexpress thing. Is there a compatible list of displays or a way to distinguish? Maybe the fact that 28-touchscreen fails to get a buffer means the screen I have isn't compatible with touch on the pi5

I'll get a usb to serial adapter and see if there's any useful logs. Failing that it sounds like I might have to drop down to a pi 4b

**Reply by Luke-Shepp** [reply]
02-screenpixel works. I'll port my code over to this sample and see at what point it breaks. A little blind debugging until I get a usb to serial, but will see where it leads.

**Reply by Luke-Shepp** [reply]
Narrowed it down to 2 places. 

A file that `#include <lvgl/lvgl.h>`, and also attempting to `m_Serial.Initialize(9600);`.

The first I presume is the limitations of framebuffer on the pi5 - looks like I'll have to go and get a 4b instead ☹️

**Reply by rsta2** [reply]
No, there isn't a display white list.

No, the message "Cannot get touch buffer" comes from the RPi touchscreen display driver and tells, that the firmware does not support it. You could comment out [this line](https://github.com/rsta2/circle/blob/749fe99301846217439a40ac7bda547a986db0f7/sample/28-touchscreen/kernel.cpp#L86), if you want to use an USB touchscreen.

LVGL works on the RPi 5, if the HDMI display works with it.

**Comment by zarkone** [comment]
I can't get circle (tic-80 baremetal rpi4 build) to work with [uConsole](https://www.clockworkpi.com/uconsole), probably because of something similar. On HDMI I see 4 squares, while built-in screen doesn't work at all. This was cm4 lite. 

On rpi 3+ and old tic-80 build I can see it works, but only HDMI, built-in screen doesn't work as well as other peripherals(keyboard, USB keyboard) 

I realized that I simply don't understand how drivers works with circle. Is it possible to use Linux drivers?

**Reply by zarkone** [reply]
Or maybe, instead of guessing, I should rather try to create a minimal circle app which displays something. Then, if it works, it would mean that something is wrong with tic-80 build

**Reply by rsta2** [reply]
uConsole may need additional drivers to work with Circle. Linux drivers cannot be used. The minimal example program in Circle to work with a HDMI display is [sample/02-screenpixel](https://github.com/rsta2/circle/tree/master/sample/02-screenpixel).

**Reply by zarkone** [reply]
Thanks! In my case, I need non-hdmi display to work, but I guess that would be a great start also.
